core/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java

```diff
@@ -121,23 +121,26 @@ public static Object parseStoredFieldsValue(XContentParser parser) throws IOExce
      * @param parser      the current {@link XContentParser}
      * @param delimiter   the delimiter to use to splits the field's name
      * @param objectClass the object class of the object to parse
+     * @param consumer    something to consume the parsed object
      * @param <T>         the type of the object to parse
-     * @return the parsed object
      * @throws IOException if anything went wrong during parsing or if the type or name cannot be derived
      *                     from the field's name
      */
-    public static <T> T parseTypedKeysObject(XContentParser parser, String delimiter, Class<T> objectClass) throws IOException {
+    public static <T> void parseTypedKeysObject(XContentParser parser, String delimiter, Class<T> objectClass, Consumer<T> consumer)
+            throws IOException {
+        assert parser.currentToken() == XContentParser.Token.START_OBJECT || parser.currentToken() == XContentParser.Token.START_ARRAY;
         String currentFieldName = parser.currentName();
         if (Strings.hasLength(currentFieldName)) {
```

Comment:
This was a defensive check in case of the method is called on anonymous object. I think there's an associated test in XContentParserUtilsTests. Now this method is more lenient we could throw an exception if currentFieldName is null I think.